### PR TITLE
Using HDF5 to store results and metadata

### PR DESCRIPTION
**Pull request by urig**
This PR changes the way Entropy stores experiment *results* and *metadata*. Instead of the default database implementation saving *results* and *metadata* in SQLite via SQLAlchemy, it now stores them in an [HDF5](https://www.hdfgroup.org/solutions/hdf5/) file.

HDF5 is a heterogeneous, hierarchical data and file format designed to store and organize large amounts of data. HDF5 offers better i/o performance over SQLite and in the near future will allow data to be easily exported from Entropy to related scientific software tools.

This new feature is subject to a feature flag ([see below](#feature_flag)) and **requires manual action for migration** from the previous version of `Entropy` (also [see below](manual_migration))

As this is our first experience with HDF5 (based on [`h5py`](https://www.h5py.org/)), data saved with `save_result()` or `save_metadata()` are now stored both in a new `entropy.hdf5` file and, as before, in the old `Results` and `ExperimentMetadata` tables in the SQLite DB. When *results* or *metadata* are read (`get_results()`, `get_metadata()` and `get_last_result_of_experiment()`) they are read exclusively from the HDF5 file.

At a later release *results* and *metadata* will be saved to and read from the HDF5 exclusively.

<a name="manual_migration"></a>
### Manual Migration
Once you've upgraded Entropy to the latest version, you will need to perform a one-time manual migration procedure. The procedure will update the structure of the SQLite database and copy previously saved *result* and *metadata* records over to the new HDF5 file.

To perform the procedure all you need to do is run the `upgrade_db()` function available in the `entropylab.results_backend.sqlalchemy` package as in this example:
```
from entropylab.results_backend.sqlalchemy import upgrade_db

upgrade_db('<path_to_your_db>')
```

<a name="feature_flag"></a>
### Feature Flag

The new HDF5 storage feature is protected by a feature flag. The feature flag is on by default. To turn the flag off, please contact the Entropy development team and we will provide you with instructions.

### HDF5 file structure

The HDF5 file is at this time internal to Entropy. We encourage users not to use the file directly and instead rely on existing API methods from `data_reader` and `data_writer`. 

Should you with to access the HDF5 file directly, please make a copy of it and work on the copy as changes to the file can have unexpected effects on Entropy.

*Results* are stored in the HDF5 file under the path `/experiments/<experiment_id>/<stage>/<label>/result` where `<experiment_id>`, `<stage>` and `<label>` are *groups* and `result` is the *dataset*. The path is therefor the unique identifier for the result. Similarly, *metadata* are stored in the HDF5 file under the path `/experiments/<experiment_id>/<stage>/<label>/metadata`.

The *dataset* contains the `ResultRecord`'/`MetadataRecord`s `data`. If `data` is a Python primitive or a [NumPy](https://numpy.org/) [`dtype`](https://numpy.org/doc/stable/user/basics.types.html) it is stored as is. In all other cases the data is stored *pickled*. If pickling is not possible, Entropy defaults to saving the `data`'s `__repr__()` as bytes (*Opaque* in HDF5 terminology).

Each *dataset* is saved with the following *attributes* (HDF5 metadata): 
- `experiment_id` 
- `stage` 
- `label` 
- `story` (only for *results*)
-  `time` ("time when saved" as [ISO 8601 string](https://docs.python.org/3/library/datetime.html#datetime.datetime.isoformat) with timezone). 

Pickled data have an additional attribute called `data_type` which represents their `ResultDataType` (`Pickled` or `String`).

### Breaking changes
* The combination of `experiment_id`, `stage` and `label` is now unique. It is no longer possible to call `save_result()` or `save_metadata()` multiple times for these same values.
* When saving experiment *results* or metadata*, `label` is no longer allowed to be `None` or an empty string and must be unique per *experiment* and *stage* combination.
* The type of the `id` of *results* and *metadata* has been changed from `int` to `str`.
* New columns are added to the `Results` and `ExperimentMetadata` tables in SQLite: `saved_in_hdf5`. This is a non-nullable `bool` column (default: `False`)
* New table added to the SQLite database: `alembic_versions`.

### CODE REVIEW

File: entropylab/results_backend/sqlalchemy/alembic/README [hunk 1]
```diff
@@ -0,0 +1 @@
+Generic single-database configuration.
```
Comment (talshani) [inline entropylab/results_backend/sqlalchemy/alembic/README:1]:
```suggestion
This alembic environment is responsible for migrating the database for the results metadata Entropy save for each run
```

Comment (talshani) [reply entropylab/results_backend/sqlalchemy/alembic/README:1]: Make the readme more informative

File: entropylab/results_backend/sqlalchemy/db.py [hunk 2]
```diff
@@ -1,416 +1,467 @@
...
+    MetadataTable,
+    NodeTable,
+)
+
+T = TypeVar(
+    "T",
+)
+
+
+# noinspection PyBroadException
```
Comment (talshani) [inline entropylab/results_backend/sqlalchemy/db.py:60]: Can we just narrow the exception type and not ignore lint errors?

File: entropylab/results_backend/sqlalchemy/model.py [hunk 3]
```diff
@@ -47,6 +47,7 @@ def _get_class(module_name, class_name):
     return cls
 
 
+# noinspection PyBroadException
```
Comment (talshani) [inline entropylab/results_backend/sqlalchemy/model.py:50]: Can we just narrow the exception type?

File: entropylab/results_backend/sqlalchemy/storage.py [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+        else:
+            return []
+
+
+class EntityType(Enum):
+    RESULT = 1
+    METADATA = 2
+
+
+# noinspection PyMethodMayBeStatic,PyBroadException
```
Comment (talshani) [inline entropylab/results_backend/sqlalchemy/storage.py:113]: Can we just narrow the exception type?

File: entropylab/results_backend/sqlalchemy/tests/test_db.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import os
+
+import pytest
+
+from config import settings
+from entropylab import SqlAlchemyDB, RawResultData
+
+
+HDF_FILENAME = "./entropy.hdf5"
```
Comment (talshani) [inline entropylab/results_backend/sqlalchemy/tests/test_db.py:9]: This creates an `entropy.hdf5` in the root of the project when we run tests, and it is not ignored.

Can you create this file in some temp directory, and add that directory to ignore?

File: pyproject.toml [hunk 6]
```diff
@@ -1,71 +1,74 @@
...
-
-[tool.black]
-target-version = ["py37"]
-line-length = 88
-
-[build-system]
-requires = ["poetry-core>=1.0.0"]
-build-backend = "poetry.core.masonry.api"
+[tool.poetry]
+name = "entropylab"
```
Comment (talshani) [inline pyproject.toml:2]: Double check the line endings, I think they are windows lineendings,
just edit this file to be linux lineendings

File: settings.toml [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# TODO: Do we want to include the settings.toml in the release?
```
Comment (talshani) [inline settings.toml:1]: the file is not included in the output package anyway.
do we need it in the final distribution?
is the user should add this file locally?

**Review by talshani — CHANGES_REQUESTED** [review]

File: settings.toml (hunk 7 above)
Comment (urig) [reply settings.toml:1]: Thanks. Right now this file is not needed. Should a user encounter problems with the new HDF5 feature, we can instruct them to create this file and put an entry in it that turns the feature toggle off. Shall I remove it from source control entirely for now?

File: pyproject.toml (hunk 6 above)
Comment (urig) [reply pyproject.toml]: @Tal Done. Can I ask you to verify? 🙏

File: entropylab/results_backend/sqlalchemy/alembic/README (hunk 1 above)
Comment (urig) [reply entropylab/results_backend/sqlalchemy/alembic/README:1]: I will expand this README to explain how to create new migrations etc tomorrow 🤞

File: entropylab/results_backend/sqlalchemy/db.py (hunk 2 above)
Comment (urig) [reply entropylab/results_backend/sqlalchemy/db.py:60]: I've removed all #noinspection directives. Narrowed down broad except clauses to `RuntimeError`

File: settings.toml (hunk 7 above)
Comment (talshani) [reply settings.toml:1]: Yes, remove it for now

File: pyproject.toml (hunk 6 above)
Comment (talshani) [reply pyproject.toml]: Worked!

File: entropylab/results_backend/sqlalchemy/db.py [hunk 8]
```diff
@@ -1,416 +1,466 @@
-import os
```
Comment (qguyk) [inline entropylab/results_backend/sqlalchemy/db.py:1]: please fix line endings here as well

**Review by qguyk — CHANGES_REQUESTED** [review]

File: settings.toml (hunk 7 above)
Comment (urig) [reply settings.toml]: Done

File: entropylab/results_backend/sqlalchemy/db.py (hunk 8 above)
Comment (urig) [reply entropylab/results_backend/sqlalchemy/db.py:1]: From what I see, the file has CRLF on main and my PR changes it to LF. Hence the extra large diff. Can I ask you to confirm and resolve?

**Review by talshani — APPROVED** [review]

**Pull request merged**